From f1cd7c4df5d9a7ac37cf8ece57f924482c64dba7 Mon Sep 17 00:00:00 2001
From: "Song, Yoong Siang" <yoong.siang.song@intel.com>
Date: Sun, 11 Apr 2021 23:19:13 +0800
Subject: [PATCH 11/33] REVERTME: net: stmmac: introduce AF_XDP ZC TX HW
 timestamps

This is a temporary solution as it uses trace_printk as a means to
log tx timestamps. Future implementation should use xdp_frame's
data_meta to let user applications retrieve it directly.

Signed-off-by: Voon Weifeng <weifeng.voon@intel.com>
Signed-off-by: Tan, Tee Min <tee.min.tan@intel.com>
Signed-off-by: Song, Yoong Siang <yoong.siang.song@intel.com>
---
 .../net/ethernet/stmicro/stmmac/stmmac_main.c  | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
index dcb39d9a77bf..19221b24b3ed 100644
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -2464,12 +2464,16 @@ static bool stmmac_xdp_xmit_zc(struct stmmac_priv *priv, u32 queue, u32 budget)
 
 		tx_q->tx_count_frames++;
 
-		if (!priv->tx_coal_frames[queue])
+		if (unlikely(priv->hwts_all)) {
+			stmmac_enable_tx_timestamp(priv, tx_desc);
+			set_ic = true;
+		} else if (!priv->tx_coal_frames[queue]) {
 			set_ic = false;
-		else if (tx_q->tx_count_frames % priv->tx_coal_frames[queue] == 0)
+		} else if (tx_q->tx_count_frames % priv->tx_coal_frames[queue] == 0) {
 			set_ic = true;
-		else
+		} else {
 			set_ic = false;
+		}
 
 		if (set_ic) {
 			tx_q->tx_count_frames = 0;
@@ -2604,6 +2608,14 @@ static int stmmac_tx_clean(struct stmmac_priv *priv, int budget, u32 queue)
 				stmmac_get_tx_hwtstamp(priv, p,
 						       &shhwtstamp.hwtstamp);
 				skb_tstamp_tx(skb, &shhwtstamp);
+			} else if (unlikely(priv->hwts_all) &&
+				   tx_q->tx_skbuff_dma[entry].buf_type ==
+				   STMMAC_TXBUF_T_XSK_TX) {
+				ktime_t tx_hwtstamp = { 0 };
+
+				stmmac_get_tx_hwtstamp(priv, p, &tx_hwtstamp);
+				trace_printk("XDP TX HW TS %llu\n",
+					     tx_hwtstamp);
 			}
 		}
 
-- 
2.32.0

